database/merkledb/root_node.go

```diff
@@ -0,0 +1,157 @@
+package merkledb
+
+import "fmt"
+
+// RootNode is the top most node of the tree
+type RootNode struct {
+	Child              []byte `serialize:"true"`
+	StoredHash         []byte `serialize:"true"`
+	RootID             uint32 `serialize:"true"`
+	previousStoredHash []byte
+	persistence        *Persistence
+}
+
+// NewRootNode returns a RootNode without children
+func NewRootNode(id uint32, p *Persistence) Node {
+	return &RootNode{RootID: id, persistence: p}
+}
+
+// GetChild returns the child
+func (r *RootNode) GetChild(key Key) (Node, error) {
+	if len(r.Child) == 0 {
+		return nil, nil
+	}
+	node, err := r.persistence.GetNodeByHash(r.Child)
+	if err != nil {
+		return nil, err
+	}
+	node.SetParent(r)
+	node.SetPersistence(r.persistence)
+	return node, nil
+}
+
+// GetNextNode returns the child if it respects the nextNode inputs
+func (r *RootNode) GetNextNode(prefix Key, start Key, key Key) (Node, error) {
+	node, err := r.persistence.GetNodeByHash(r.Child)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(key) != 0 {
+		if !Greater(node.Key(), key) || node.Key().Equals(key) {
+			return nil, nil
+		}
+	}
+
+	if len(start) != 0 {
+		if !Greater(node.Key(), start) || node.Key().Equals(start) {
+			return nil, nil
+		}
+	}
+
+	if len(prefix) != 0 {
+		if !IsPrefixed(prefix, node.Key()) {
+			return nil, nil
+		}
+	}
```

Comment:
Yeah, good spot - moved that to a variable.

Added the following comment
```
// GetNextNode returns the child if it respects the nextNode inputs
// Instead of checking all the positive condition
// the single child Node will only be returned if it does not fail any of the negative condition
// ie. if there's a prefix it will fail/return nil, if the child Node key does not respect that condition
```

👍 👎  ?